sql/postgresql/pglexer.g4

```diff
@@ -0,0 +1,1045 @@
+/* This field stores the tags which are used to detect the end of a dollar-quoted string literal.
+ */
+private readonly Stack<String> _tags = new Stack<String>();
+}
+
+//
+// SPECIAL CHARACTERS (4.1.4)
+//
+
+// Note that Asterisk is a valid operator, but does not have the type Operator due to its syntactic use in locations
+// that are not expressions.
+
+Dollar			: '$';
+OPEN_PAREN		: '(';
+CLOSE_PAREN		: ')';
+OPEN_BRACKET		: '[';
+CLOSE_BRACKET	: ']';
+COMMA			: ',';
+SEMI    		: ';';
+COLON			: ':';
+STAR     		: '*';
+EQUAL			: '=';
+DOT			: '.';
+//NamedArgument	: ':=';
+PLUS            : '+';
+MINUS           : '-';
+LT              : '<';
+GT              : '>';
+LESS_LESS       : '<<';
+GREATER_GREATER : '>>';
+COLON_EQUALS    : ':=';
+LESS_EQUALS     : '<=';
+EQUALS_GREATER  : '=>';
+GREATER_EQUALS  : '>=';
+DOT_DOT         : '..';
+NOT_EQUALS                           : '<>';
+TYPECAST            : '::';
+PERCENT : '%';
+PARAM : '$'([0-9])+;
+
+
+//
+// OPERATORS (�4.1.3)
+//
+
+// this rule does not allow + or - at the end of a multi-character operator
+Operator
+	:	((	OperatorCharacter
+		|	(	'+'
+			|	'-' {InputStream.LA(1) != '-'}?
+			)+
+			(	OperatorCharacter
+			|	'/' {InputStream.LA(1) != '*'}?
+			)
+		|	'/' {InputStream.LA(1) != '*'}?
+		)+
+	|	// special handling for the single-character operators + and -
+		[+-]
+		)
+		//TODO somehow rewrite this part without using Actions
+		{
+		    if(Text=="<<") Type = LESS_LESS;
+		    if(Text==">>") Type = GREATER_GREATER;
+		}
+	;
+ 
+/* This rule handles operators which end with + or -, and sets the token type to Operator. It is comprised of four
+ * parts, in order:
+ *
+ *   1. A prefix, which does not contain a character from the required set which allows + or - to appear at the end of
+ *      the operator.
+ *   2. A character from the required set which allows + or - to appear at the end of the operator.
+ *   3. An optional sub-token which takes the form of an operator which does not include a + or - at the end of the
+ *      sub-token.
+ *   4. A suffix sequence of + and - characters.
+ */
+OperatorEndingWithPlusMinus
+	:	(	OperatorCharacterNotAllowPlusMinusAtEnd
+		|	'-' {InputStream.LA(1) != '-'}?
+		|	'/' {InputStream.LA(1) != '*'}?
+		)*
+		OperatorCharacterAllowPlusMinusAtEnd
+		Operator?
+		(	'+'
+		|	'-' {InputStream.LA(1) != '-'}?
+		)+
+		-> type(Operator)
+	;
+
+// Each of the following fragment rules omits the +, -, and / characters, which must always be handled in a special way
+// by the operator rules above.
+
+
+fragment
+OperatorCharacter
+	: [*<>=~!@%^&|`?#]
+	;
+
+// these are the operator characters that don't count towards one ending with + or -
+fragment
+OperatorCharacterNotAllowPlusMinusAtEnd
+	:	[*<>=+]
+	;
+
+// an operator may end with + or - if it contains one of these characters
+fragment
+OperatorCharacterAllowPlusMinusAtEnd
+	:	[~!@%^&|`?#]
+	;
+
+//
+// KEYWORDS (Appendix C)
+//
+
+//
+// reserved keywords
+//
+ALL					: [Aa][Ll][Ll];
```

Comment:
I suggest using fragment rules for clarity:

```suggestion
ALL					: A L L;
...
fragment A : [Aa];
fragment L : [Ll];
```

Also, [Case insensitive stream](https://github.com/antlr/antlr4/blob/master/doc/case-insensitive-lexing.md) allows writing lexer rules in the simple style:

```suggestion
ALL					: 'ALL';
```